File: IanLenehan/week_01/Friday/MTA.js

var lines = {
  'L line': ["8th", "6th", "Union Square", "3rd", "1st"],
  'N line': ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"],
  '6 line': ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"],
}

// determine if the line exists

var lineExists = function (line) {
  if (!(line in lines)) {
  //  console.log(line + ' exists!');
    return false;
  } else {
    return true;
  }
}

// determine if the line and station exist

var stationExists = function (station, line) {
  var lineTrue = lineExists(line);
  if (lineTrue === false) {
    console.log(line + " does not exist!");
    return false;
  } else {

  var matches = 0;
  for (var i = 0; i < lines[line].length; i++) {
      if (lines[line][i] === station) {
        //console.log("exists");
        matches = 1;
        return true;
      }
  }
if (matches < 1) {
  //console.log(station + " doesn't " + line);
  return false;
}
}
}



// lineExists("Q Line");

//first test is to create an array of the stations traveled for each leg
//cycle through the specific array to find the stops between the indexes
//and add them to the array

var planTripBeta = function(fromStop, line, toStop) {
  var fromIndex = lines[line].indexOf(fromStop);
  var toIndex = lines[line].indexOf(toStop);
  var journey = [];
  var message = '';
  if(fromIndex > toIndex) {
    // if the first station is further on the track than the next stations
    // reverse the array
    for (var i = toIndex + 1; i < fromIndex; i++) {
      journey.push(lines[line][i]);
      journey = journey.reverse();
    }
  } else {
    for (var i = fromIndex + 1; i < toIndex; i++) {
      journey.push(lines[line][i])

    }
  }
  //console.log(journey);
  return journey.join( " station and " );

}

// calculate the distance between each stop for each leg of the journey

var tripDistance = function (fromStop, line, toStop) {
  var fromIndex = lines[line].indexOf(fromStop);
  var toIndex = lines[line].indexOf(toStop);
  if(fromIndex > toIndex) {
    var distance = fromIndex - toIndex;
  } else {
    distance = toIndex - fromIndex;
  }
  //console.log(distance);
  return distance;
}

// tripDistance("34th", "N line", "8th");


var planTrip = function(fromStop, fromLine, toStop, toLine) {

  var fromStopExists = stationExists(fromStop, fromLine);
  var toStopExists = stationExists(toStop, toLine);

// first check if both stops exist

if (fromStopExists === false) {
    console.log(fromStop + " does not exist on the " + fromLine + "!");
    return false;
  } else if (toStopExists === false) {
    console.log(toStop + " does not exist on the " + toLine + "!");
    return false;
  } else {

    var fromIndex = lines[fromLine].indexOf(fromStop);
    var toIndex = lines[toLine].indexOf(toStop);
    var distance1 = '';
    var distance2 = '';
    var journey1 = [];
    var journey2 = [];
    var message = '';

    if (fromLine !== toLine) {
      journey1 = planTripBeta(fromStop, fromLine, "Union Square");
      distance1 = tripDistance(fromStop, fromLine, "Union Square");
      journey2 = planTripBeta("Union Square", toLine, toStop);
      distance2 = tripDistance("Union Square", toLine, toStop);
      var message = ("Starting at " + fromStop + " station on the " + fromLine + ", travel " +
      distance1 + " stops, through " + journey1 + " station to Union Square station, where you will change to the "
      + toLine + ". You will travel " + distance2 + " more stops, through " + journey2 +
       " station into " + toStop) + " station.";
      //console.log(journey1, distance1, journey2, distance2);
    } else {
      journey1 = planTripBeta(fromStop, fromLine, toStop);
      distance1 = tripDistance(fromStop, fromLine, toStop);
      if (fromStop !== toStop) {
      message = ("Starting at " + fromStop + " station on the " + fromLine + ", travel " +
      distance1 + " stops to " + toStop) + " station.";
    } else {
      message = ("Pay attention!");
    }
    }
    console.log(message);
    //console.log(message);
    return message;

  }

}



planTrip("34th", "N line", "28th", "6 line");
planTrip("8th", "N line", "8th", "L line");
planTrip("23rd", "N line", "Times Square", "N line");

planTrip("Times Square", "N line", "Grand Central", "6 line"); //longest possible journey in one direction
planTrip("Grand Central", "6 line", "Times Square", "N line"); //longest possible journey in other direction

planTrip("Grand Piano", "6 line", "Times Square", "N line"); //with a station that doesn't exist
planTrip("Grand Central", "6 line", "NY Times", "N line"); //with another station that doesn't exist
planTrip("Grand Central", "Q line", "Times Square", "N line"); //with a line that doesn't exist

planTrip("Grand Central", "6 line", "Grand Central", "6 line"); //Same to and from station.

File: IanLenehan/week_04/ruby_MTA/mta.rb
LINES = {
  "N Line" => ["Times Square", "34th", "28th", "23rd", "Union Square", "8th"],
  "L Line" => ["8th", "6th", "Union Square", "3rd", "1st"],
  "6 Line" => ["Grand Central", "33rd", "28th", "23rd", "Union Square", "Astor Place"]
}

print "Please choose a line: "
@from_line = gets.chomp
print "From station: "
@from_stop = gets.chomp
print "To Line: "
@to_line = gets.chomp
print "To station: "
@to_stop = gets.chomp
@trip_array = []

# check if the line exists by lookng into the LINES hash to see if there is a key with the same name
def line_exists(line)
  LINES.key?(line)
end

# check the stop exists on that line by checking to see if that array includes a stop by that name
def stop_exists(stop, line)
  LINES[line].include? stop
end

# create a function that evaluates each leg separately
def single_trip_array (from_stop, line, to_stop)
  # find the positions/index of each stop on the line
  from_index = LINES[line].index(from_stop)
  to_index = LINES[line].index(to_stop)
  # determine the direction of the array
  if from_index < to_index
    @trip_array = LINES[line][from_index..to_index]
    # remove the last item from the array, as it will be the same as to_stop
    # I could also use ... instead of .. as an exclusive array
    @trip_array.pop
    @trip_array.shift # removes the first item as that is the from_stop
  else
    @trip_array = LINES[line][to_index..from_index].reverse
    @trip_array.pop
    @trip_array.shift
  end
  @trip_array
end

# calculate distance between indexes
def single_trip_distance (from_stop, line, to_stop)
  from_index = LINES[line].index(from_stop)
  to_index = LINES[line].index(to_stop)
  distance = (to_index - from_index)
  distance.abs # converts any negative integer to a positive
end

def plan_trip
  if line_exists(@from_line) == false
    p "#{@from_line} does not exist!"
    return
  elsif line_exists(@to_line) == false
    p "#{@to_line} does not exist!"
    return
  elsif stop_exists(@from_stop, @from_line) == false
    p "There is no #{@from_stop} on the #{@from_line}!"
    return
  elsif stop_exists(@to_stop, @to_line) == false
    p "There is no #{@to_stop} on the #{@to_line}!"
  end
  if @from_stop == @to_stop
    p "Surely you can walk?"
    return
  end
  if @from_line == @to_line
    # if the line is the same, simply run each method once
    distance = single_trip_distance(@from_stop, @from_line, @to_stop)
    stops = single_trip_array(@from_stop, @from_line, @to_stop)
    message = "Board the #{@from_line} at #{@from_stop} station, travel #{distance} stops through #{stops.join(", ")} until you arrive at #{@to_stop} station."
  else
    # if we're running two lines, to_stop on trip one and from_stop on trip two is always Union Square
    distance1 = single_trip_distance(@from_stop, @from_line, "Union Square")
    stops1 = single_trip_array(@from_stop, @from_line, "Union Square")
    distance2 = single_trip_distance("Union Square", @to_line, @to_stop)
    stops2 = single_trip_array("Union Square", @to_line, @to_stop)
    message = "Board the #{@from_line} at #{@from_stop} station, travel #{distance1} stops through #{stops1.join(", ")} into Union Square. Get off here and jump on the #{@to_line}. Travel #{distance2} more stops through #{stops2.join(", ")} into #{@to_stop} station."
  end
  p message
end

plan_trip
#
# require 'pry'
# binding.pry
